### PR TITLE
AC-532 Adding validation to fields of contact information 

### PR DESCRIPTION
**Pull request by Asgardian8740**
AC-532 Adding validation to fields of contact information 

   

## Description of what I changed
     In this commit I have added validation to contact information fields in the Register patient activity.
    

## Issue I worked on

JIRA Issue: https://issues.openmrs.org/browse/AC-532

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 1]
```diff
@@ -451,6 +485,11 @@ private void resolveViews(View v) {
         gendererror = (TextView)v.findViewById(R.id.gendererror);
         addrerror = (TextView)v.findViewById(R.id.addrerror);
         countryerror = (TextView)v.findViewById(R.id.countryerror);
+        stateerror=(TextView)v.findViewById(R.id.stateerror);
+        cityerror=(TextView)v.findViewById(R.id.cityerror);
+        countrynull=(TextView)v.findViewById(R.id.countrynull);
+        postalerror=(TextView)v.findViewById(R.id.postal_error);
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @Asgardian8740 Can you format this code with Android Studio?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 2]
```diff
@@ -222,6 +231,30 @@ public void setErrorsVisibility(boolean givenNameError,
...
+        }
+        else {
+            cityerror.setVisibility(View.GONE);
+        }
+        if (postalError) {
+            postalerror.setVisibility(View.VISIBLE);
+        }
+        else {
+            postalerror.setVisibility(View.GONE);
+        }
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @Asgardian8740 Turn this into a function: one that sets a view to either visible or gone if its corresponding error is true.

Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: Then replace all these if else statements with the function call you just created.

File: openmrs-client/src/main/res/layout/fragment_patient_info.xml [hunk 3]
```diff
@@ -111,8 +114,8 @@
                     android:layout_height="wrap_content"
                     android:layout_marginLeft="20dp"
                     android:layout_marginRight="20dp"
-                    android:weightSum="2"
-                    android:baselineAligned="false">
+                    android:baselineAligned="false"
+                    android:weightSum="2">
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/layout/fragment_patient_info.xml]: @Asgardian8740 Please reformat this xml layout file with Android Studio.

File: openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/AddEditPatientPresenterTest.java [hunk 4]
```diff
@@ -191,7 +191,7 @@ public void shouldUpdatePatient_errorResponse(){
                 .thenReturn(mockSuccessCall(new PatientPhoto()));
 
         presenter.confirmUpdate(patient);
-        verify(view).setErrorsVisibility(false, false, false, false, false, false);
+        verify(view).setErrorsVisibility(false, false, false, false, false, false,false,false,false,false);
```
Comment (f4ww4z) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/AddEditPatientPresenterTest.java]: @Asgardian8740 Reformat this file with Android Studio's default formatter.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java [hunk 5]
```diff
@@ -222,6 +231,30 @@ public void setErrorsVisibility(boolean givenNameError,
         else {
             gendererror.setVisibility(View.GONE);
         }
```
Comment (Asgardian8740) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @f4ww4z  Pardon,I didn't understand the change which you asked and also do i have to make this changes for the rest of the code in the given function or only to the code I added?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java (hunk 2 above)
Comment (Asgardian8740) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @f4ww4z Pardon,I didn't understand the change which you asked and also do i have to make this changes for the rest of the code in the given function or only to the code I added?

Comment (f4ww4z) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java]: @Asgardian8740 You can turn the `if ... else` block into a method that takes a TextView (error message) and boolean (if there's an error) as an argument. Then set the visibility based on the argument. Yes, ideally call this method in as much places as possible.

**Comment by Rimjhim28** [comment]
@f4ww4z Can you please check the merge again. I guess it's causing AddEditPatientActivity to crash.
The error occurs when I try to register a new patient.

**Comment by Asgardian8740** [comment]
> @f4ww4z Can you please check the merge again. I guess it's causing AddEditPatientActivity to crash.
> The error occurs when I try to register a new patient.

@Rimjhim28  Mam,This error is due to "Error inflating class android.support.design.widget.FloatingActionButton" which is most probably because of issue due to design library version.
I guess this is because of changes after merging AC-535.

@f4ww4z  May I raise this issue in JIRA and start working on it?

**Comment by deepak140596** [comment]
Hi, it was happening because as the project was Migrated to AndroidX on AC-535, the components in the form weren't updated to the same. They were using the old support library. 
As I was working on the same Activity on ticket AC-539, I have already done those necessary changes.

As soon as the PR is reviewed and merged, the error will be resolved.

**Comment by deepak140596** [comment]
Also as per the code of the Address Validation, we do not require to get the whole address of the patient. We just need atleast one field to be filled. Therefore, adding validation error fields in all the address' fields do not make sense. 
@f4ww4z please correct me if I am wrong.

**Comment by Asgardian8740** [comment]
> Hi, it was happening because as the project was Migrated to AndroidX on AC-535, the components in the form weren't updated to the same. They were using the old support library.
> As I was working on the same Activity on ticket AC-539, I have already done those necessary changes.
> 
> As soon as the PR is reviewed and merged, the error will be resolved.

ohk thats great!!

**Comment by Asgardian8740** [comment]
> Also as per the code of the Address Validation, we do not require to get the whole address of the patient. We just need atleast one field to be filled. Therefore, adding validation error fields in all the address' fields do not make sense.
> @f4ww4z please correct me if I am wrong.

See I guess the purpose of the address is to have the contact information of the patient.This is the only information we are taking from the patient and hence in the time of need,we cannot contact him/her with only country or city or state or pincode or street address.Hence to remove ambiguity in communication we require complete address.

**Pull request merged**